### ISSUE TITLE
bounded memory check fails

### ISSUE THREAD

**Issue by nrainer-materialize** [MaterializeInc/materialize#23175]
### Buildkite link

https://buildkite.com/materialize/nightlies/builds/5183#018bcb3c-a5db-40cc-983c-c51073d28a32

### Relevant log output

```source
> UPDATE t SET a = a + 1000000;
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 49s 74s 111s 166s 249s 374s 561s 842s 964s
^^^ +++
15:1: error: executing query failed: db error: ERROR: canceling statement due to statement timeout
HINT: Consider increasing the maximum allowed statement duration for this session by setting the statement_timeout session variable. For example, `SET statement_timeout = '60s'`.: ERROR: canceling statement due to statement timeout
HINT: Consider increasing the maximum allowed statement duration for this session by setting the statement_timeout session variable. For example, `SET statement_timeout = '60s'`.
     |
  14 | > UPDATE t SET a = a + 100000;
  15 | > UPDATE t SET a = a + 1000000;
     | ^
+++ !!! Error Report
1 errors were encountered during execution
```

**Comment by nrainer-materialize** [comment]
Trying to bisect

**Comment by nrainer-materialize** [comment]
Assuming to be caused by https://github.com/MaterializeInc/materialize/pull/23141. Still double-checking...

**Comment by nrainer-materialize** [comment]
This is caused by https://github.com/MaterializeInc/materialize/pull/22792, which sets `compute_dataflow_max_inflight_bytes` to `134217728` in CI. (It was flaky before but occurs every time since this change.)

Reproducible in scratch with `bin/mzcompose --find bounded-memory down -v && bin/mzcompose --find bounded-memory run default table-index-hydration`.

@danhhz: Can you have a look at it? How to proceed with this flag?

(Note that https://github.com/MaterializeInc/materialize/pull/23217 will change the setup.)

**Comment by danhhz** [comment]
Given that this is behind an LD flag, and the bisected PR is just flipping it on in CI, I'm going to remove the release blocker label from this

**Comment by nrainer-materialize** [comment]
Alright, things are as follows...

**TLDR:**
* The scenario `pg-cdc-snapshot` failed at the beginning of the week.
* Configuring `compute_dataflow_max_inflight_bytes` in CI caused `pg-cdc-snapshot` to succeed but `table-index-hydration` to fail. Overall, the flag is expected to help memory bounded.
* `pg-cdc-snapshot` is a regression. `table-index-hydration` performs worse with the flag set.

**More details:**
* Two scenarios are mainly relevant: `pg-cdc-snapshot` and `table-index-hydration` 
* The flag `compute_dataflow_max_inflight_bytes` to `134217728` (https://github.com/MaterializeInc/materialize/pull/22792 and merged with 5eb580c12501fec0e1b508613665503efb993092) has positive and negative consequences.
* This flag configuration caused the scenario `table-index-hydration` to fail (and `pg-cdc-snapshot` to succeed).
* Reverting the flag (https://github.com/MaterializeInc/materialize/pull/23220 merged with https://github.com/MaterializeInc/materialize/commit/cac0d3f8aa6be8e62340e3934f788f9c771c2703) caused `pg-cdc-snapshot` to fail (again) and `table-index-hydration` to succeed. `pg-cdc-snapshot` will even fail with 1.5Gb additional clusterd memory.
* Citing @danhhz [(Slack)](https://materializeinc.slack.com/archives/CTESPM7FU/p1700062374364079?thread_ts=1700036635.289049&cid=CTESPM7FU): "The irony here is that this setting is intended to help keep memory bounded and as of last night, we have [pretty definitive proof](https://materializeinc.slack.com/archives/C053EPHMU05/p1700007307689559) that it works."

**Next actions:**
* @danhhz, I suggest to reintroduce the `compute_dataflow_max_inflight_bytes` flag (reverting the revert in https://github.com/MaterializeInc/materialize/pull/23220). I will prepare a PR.
* I will try to find a configuration in which `table-index-hydration` succeeds with the flag set.
* I will further analyze when `pg-cdc-snapshot` first failed.

See also: discussion in https://materializeinc.slack.com/archives/CTESPM7FU/p1700036635289049

**Comment by nrainer-materialize** [comment]
### Update
I have not been able to identify the commit that caused `pg-cdc-snapshot` to fail with my EC2 setup; if may be one or multiple commits. The main challenge is that my EC2 instance likes to fail more often; even on older builds where nightly was green. In addition, results are not 100% stable (but not too bad).
I updated the raw results above with data from further builds.

As for the search for a working configuration of table-index-hydration, I switched back to the CI and will try the commits of https://github.com/MaterializeInc/materialize/pull/23273/commits.

**Comment by danhhz** [comment]
Rainer, going to assign this to you since it seems like you're looking into it (and I am not :) )

**Comment by philip-stoev** [comment]
PR with a fix: https://github.com/MaterializeInc/materialize/pull/23306

It was a problem with the test , as follows:

1. Previously pg-cdc-snapshot was failing , which was fixed by the new compute_dataflow_max_inflight_bytes setting.
2. This caused the DML statemennts to slow down, as they construct dataflows in order to obtain a list of records to insert/update
3. This caused the "canceling statement due to statement timeout" message to start showing up. It was **NOT** a sign of an OOM, it is just that the statement genuinely could no longer complete within the timeout.
4. Testdrive kept retrying those statements for up to the full 1 hour of its own timeout.

So, with the PR above, the entire bounded-memory scenario is now green.
